src/mlpack/core/optimizers/cne/cne.hpp

```diff
@@ -0,0 +1,220 @@
+ * for the next generation and evolution takes place.
+ *
+ * The evolution process basically takes place in two types-
+ * Crossover
+ * Mutation 
+ *
+ * Crossover takes two parents and generates two childs from them. Both the
+ * child have properties inherited by their parents. The parameters of the
+ * parents gets mixed using equal probability selection, creating two childs.
+ * This is just a mix of link weights or the parameters.
+ *
+ * In mutation parameters are updated by pertubating small noise.
+ * If $$ \Lambda $$ is the number of weights in the network.
+ *
+ * $$ \sum_{n=1}^{\Lambda}\omega_n = \omega_n + \rho $$
+ *
+ * where $$ \rho $$ is the small pertubation value determined randomly between
+ * 0 and the mutation size given by the user as a constructor parameter. Also
+ * the mutation probability taken as a constructor parameter decides the amount
+ * of mutation addition into the network.
+ *
+ * Both the above mentioned process creates new candidates as well as changes
+ * the existing candidates to obtain better candidates in the next generated.
+ *
+ * The whole process then repeats for multiple generation until atleast one of 
+ * the termination criteria is met-
+ *
+ * 1) The final value of the objective function (Not considered if not provided)
+ * 2) The maximum number of generation reached (optional but highly recommended)
+ * 3) Minimum change in best fitness values between two consecutive generations 
+ *    should be greater than a threshold value (Not considered if not provided) 
+ *
+ * After which the final value and the parameters are returned by the optimize
+ * method.
+ *
+ * For CNE to work, a DecomposableFunctionType template parameter is required.
+ * This class must implement the following function:
+ *
+ * size_t NumFunctions();
+ * 
+ * NumFunctions() should return the number of functions or data points in
+ * the dataset.
+ *
+ * double Evaluate(const arma::mat& coordinates, const size_t i);
+ *
+ * The parameter i refers to which individual function is being evaluated.
+ * The function will evaluate the objective function on the data point i in
+ * the dataset.
+ */
+class CNE
+{
+ public:
+/**
+ * Constructor for the CNE optimizer.
+ * 
+ * The default values provided over here are not necessarily suitable for a
+ * given function. Therefore it is highly recommended to adjust the
+ * parameters according to the problem.
+ *
+ * @param populationSize The number of candidates in the population
+ * @param maxGenerations The maximum number of generations allowed for CNE
+ * @param mutationProb   Probability that a weight will get mutated 
+ * @param mutationSize   The range of mutation noise to be added. This range
+ *                       is between 0 and mutationSize
+ * @param selectPercent  The percentage of candidates to select to become the
+ *                       the next generation
+ * @param finalValue     The final value of the objective function for  
```

Comment:
Do you think we could rename the parameter to tolerance? That way we could use the same name as we did for SGD.